core/src/consensus.rs

```diff
@@ -637,36 +733,390 @@ impl Tower {
         bank_weight
     }
 
-    fn initialize_lockouts_from_bank_forks(
+    fn voted_slots(&self) -> Vec<Slot> {
+        self.lockouts
+            .votes
+            .iter()
+            .map(|lockout| lockout.slot)
+            .collect()
+    }
+
+    pub fn is_stray_last_vote(&self) -> bool {
+        if let Some(last_voted_slot) = self.last_voted_slot() {
+            if let Some(stray_restored_slot) = self.stray_restored_slot {
+                return stray_restored_slot == last_voted_slot;
+            }
+        }
+
+        false
+    }
+
+    // The tower root can be older/newer if the validator booted from a newer/older snapshot, so
+    // tower lockouts may need adjustment
+    pub fn adjust_lockouts_after_replay(
+        self,
+        replayed_root_slot: Slot,
+        slot_history: &SlotHistory,
+    ) -> Result<Self> {
+        info!(
+            "adjusting lockouts (after replay up to {}): {:?}",
+            replayed_root_slot,
+            self.voted_slots()
+        );
+
+        assert_eq!(slot_history.check(replayed_root_slot), Check::Found);
+        // reconcile_blockstore_roots_with_tower() should already have aligned these.
+        assert!(self.root().is_some());
```

Comment:
hmmm maybe I don't think I see anything in `reconcile_blockstore_roots_with_tower ` here: https://github.com/solana-labs/solana/pull/10718/files#diff-d848cba681535238849e7e33ed32702bR1096 that guarantees `self.root().is_some()`, the restored tower seems like it could have no root if the validator never set a root before restart?

Reply:
The comment `// reconcile_blocks ...` is moved a bit to clarify what it's describing: 0bfb8227e04077e10ae5107d47569fccefaba648

Also, I've changed the handling of `root` of `Tower` to always be `Some` of `Slot`. Added some justification here: 0bfb8227e04077e10ae5107d47569fccefaba648

Reply:
makes sense, thanks!